Sources/XCTest/XCTestMain.swift

```diff
@@ -88,64 +63,30 @@ internal struct XCTRun {
     //   "All tests".
     // - An `XCTestSuite` representing the .xctest test bundle is not included.
     let selectedTestName = ArgumentParser().selectedTestName
-    var rootTestSuites = [XCTestSuite]()
+    let rootTestSuite: XCTestSuite
+    let currentTestSuite: XCTestSuite
     if selectedTestName == nil {
-        rootTestSuites.append(XCTestSuite(name: "All tests"))
-        rootTestSuites.append(XCTestSuite(name: "\(testBundle.bundlePath.lastPathComponent).xctest"))
+        rootTestSuite = XCTestSuite(name: "All tests")
+        currentTestSuite = XCTestSuite(name: "\(testBundle.bundlePath.lastPathComponent).xctest")
+        rootTestSuite.addTest(currentTestSuite)
     } else {
-        rootTestSuites.append(XCTestSuite(name: "Selected tests"))
+        rootTestSuite = XCTestSuite(name: "Selected tests")
+        currentTestSuite = rootTestSuite
     }
 
     let filter = TestFiltering(selectedTestName: selectedTestName)
-    let filteredTestCases = TestFiltering.filterTests(testCases, filter: filter.selectedTestFilter)
-
-    // When `XCTestSuite` objects are announced, they need to already include
-    // the correct tests.
-    for (testCase, _) in filteredTestCases {
-        let testCaseSuite = XCTestSuite(name: "\(testCase.init().dynamicType)")
-        rootTestSuites.last!.addTest(testCaseSuite)
-    }
-
-    for suite in rootTestSuites {
-        observationCenter.testSuiteWillStart(suite)
-    }
-
-    let overallDuration = measureTimeExecutingBlock {
-        // FIXME: This was the simplest implementation that didn't involve
-        //        changing how swift-corelibs-xctest builds up and executes a
-        //        collection of test cases. However, instead of using an index
-        //        to enumerate both the `XCTestSuite` and the test cases at
-        //        once, we should enumerate the test cases in the `XCTestSuite`.
-        //        `XCTestSuite` should be responsible for representing the
-        //        collection of tests we execute here.
-        for (index, test) in filteredTestCases.enumerated() {
-            let (testCase, tests) = test
-            let testSuite = rootTestSuites.last!.tests[index] as! XCTestSuite
-            observationCenter.testSuiteWillStart(testSuite)
-            testCase.invokeTests(tests)
-            observationCenter.testSuiteDidFinish(testSuite)
+    for (testCaseType, tests) in TestFiltering.filterTests(testCases, filter: filter.selectedTestFilter) {
+        let dummyTestCase = testCaseType.init(name: "None") { _ in }
+        let testCaseSuite = XCTestSuite(name: "\(dummyTestCase.dynamicType)")
```

Comment:
the test and is used by all test
> +    /// assertions.
> +    /// - Parameter description: The description of the failure being reported.
> +    /// - Parameter filePath: The file path to the source file where the failure
> +    ///   being reported was encountered.
> +    /// - Parameter lineNumber: The line number in the source file at filePath
> +    ///   where the failure being reported was encountered.
> +    /// - Parameter expected: `true` if the failure being reported was the
> +    ///   result of a failed assertion, `false` if it was the


2016-04-03 21:28 GMT+02:00 Brian Croom <notifications@github.com>:

> In Sources/XCTest/XCTestMain.swift
> <https://github.com/apple/swift-corelibs-xctest/pull/86#discussion_r58316019>
> :
>
> > -        // FIXME: This was the simplest implementation that didn't involve
> > -        //        changing how swift-corelibs-xctest builds up and executes a
> > -        //        collection of test cases. However, instead of using an index
> > -        //        to enumerate both the `XCTestSuite` and the test cases at
> > -        //        once, we should enumerate the test cases in the `XCTestSuite`.
> > -        //        `XCTestSuite` should be responsible for representing the
> > -        //        collection of tests we execute here.
> > -        for (index, test) in filteredTestCases.enumerated() {
> > -            let (testCase, tests) = test
> > -            let testSuite = rootTestSuites.last!.tests[index] as! XCTestSuite
> > -            observationCenter.testSuiteWillStart(testSuite)
> > -            testCase.invokeTests(tests)
> > -            observationCenter.testSuiteDidFinish(testSuite)
> > +    for (testCaseType, tests) in TestFiltering.filterTests(testCases, filter: filter.selectedTestFilter) {
> > +        let dummyTestCase = testCaseType.init(name: "None") { _ in }
> > +        let testCaseSuite = XCTestSuite(name: "\(dummyTestCase.dynamicType)")
>
> I think we could skip creating the dummy test case and just stringify the
> testCaseType value itself here?
>
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
> <https://github.com/apple/swift-corelibs-xctest/pull/86/files/651fd407f55f25eba400bf19003553f236e22083#r58316019>
>